core/services/vrf/listener_v2.go

```diff
@@ -306,11 +348,234 @@ func (a fulfilledReqV2) Compare(b heaps.Item) int {
+			// isn't.
+			reqBlockNumber := new(big.Int).SetUint64(vrfRequest.Raw.BlockNumber)
+			currBlock := new(big.Int).SetUint64(lsn.getLatestHead())
+			m := bigmath.Max(reqBlockNumber, currBlock)
+			callback, errCb := lsn.coordinator.GetCommitment(&bind.CallOpts{
+				BlockNumber: m,
+			}, vrfRequest.RequestId)
+			if errCb != nil {
+				rlog.Errorw("Unable to check if already fulfilled, processing anyways", "err", errCb)
+			} else if utils.IsEmpty(callback[:]) {
+				// If seedAndBlockNumber is zero then the response has been fulfilled
+				// and we should skip it
+				rlog.Infow("Request already fulfilled", "callback", callback)
+				lsn.markLogAsConsumed(req.lb)
+				processed[vrfRequest.RequestId.String()] = struct{}{}
+				continue
+			}
+			// Run the pipeline to determine the max link that could be billed at maxGasPrice.
+			// The ethcall will error if there is currently insufficient balance onchain.
+			maxLink, run, _, gaslimit, pf, rc, errG := lsn.getMaxLinkForFulfillment(maxGasPriceWei, req, rlog)
+			if errG != nil {
+				rlog.Warnw("Unable to get max link for fulfillment, skipping request", "err", errG)
+				continue
+			}
+			if startBalanceNoReserveLink.Cmp(maxLink) < 0 {
+				// Insufficient funds, have to wait for a user top up
+				// leave it unprocessed for now
+				rlog.Infow("Insufficient link balance to fulfill a request, breaking", "maxLink", maxLink)
+				break
+			}
+
+			// check if we've exceeded the block gas limit.
+			// if we have, we should break here and re-process this request in the next
+			// batch.
+			if (totalGasLimit + gaslimit) > blockGasLimit {
+				break
+			}
+
+			proofs = append(proofs, batch_vrf_coordinator_v2.VRFTypesProof(pf))
+			commitments = append(commitments, batch_vrf_coordinator_v2.VRFTypesRequestCommitment(rc))
+			runs = append(runs, run)
+			reqIDs = append(reqIDs, vrfRequest.RequestId)
+			lbs = append(lbs, req.lb)
+			maxLinks = append(maxLinks, maxLink)
+			totalGasLimit += gaslimit
+
+			startBalanceNoReserveLink.Sub(startBalanceNoReserveLink, maxLink)
+		}
+
+		// If we didn't enqueue anything into any of the slices
+		// then we broke out of the above loop due to out of funds.
+		// Break out of this outer loop since there's nothing left to do.
+		if len(proofs) == 0 {
+			lggr.Info("Nothing left to process in batch")
+			break
+		}
+
+		// Enqueue a single batch tx for requests that we're able to fulfill based on whether
+		// they passed simulation or not.
+		payload, err := batchCoordinatorV2ABI.Pack("fulfillRandomWords", proofs, commitments)
+		if err != nil {
+			// should never happen
+			lggr.Errorw("Failed to pack batch fulfillRandomWords payload",
+				"err", err, "proofs", proofs, "commitments", commitments)
+			continue
+		}
+
+		// Bump the total gas limit by a bit so that we account for the overhead of the batch
+		// contract's calling.
+		totalGasLimitBumped := uint64(float64(totalGasLimit) * 1.15)
```

Comment:
Simulation could underestimate because it depends on the gas used by the individual callbacks right? I think this might need to be a worst-case limit where we assume callbacks use the max

Reply:
Yeah I don't like the magic multiplier that much either. I noticed though that sometimes batch fulfillments would fail due to running out of gas, and just put this multiplier in there. What about a fresh EstimateGas on the batch fulfillment call and use a multiplier there? Or would that be slightly worse estimate than a simulation?

Reply:
yeah simulation could underestimate @Nic0S (so we'd still need a multiplier) but it does have the benefit of not relying on any relationship between totalGasLimit and the implementation of the batch fulfillment contract.  A downside of using the max callback limits is then you need to assume some upper bound per verification, whereas the simulation will deterministically tell you the verification costs. @makramkd yeah I just mean EstimateGas (which simulates under the hood)